README.md
# Hui 辉 [![Build Status](https://api.travis-ci.com/boonious/hui.svg?branch=master)](https://travis-ci.com/github/boonious/hui) [![Hex pm](http://img.shields.io/hexpm/v/hui.svg?style=flat)](https://hex.pm/packages/hui) [![Coverage Status](https://coveralls.io/repos/github/boonious/hui/badge.svg?branch=master)](https://coveralls.io/github/boonious/hui?branch=master) [![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/hui/) [![Total Download](https://img.shields.io/hexpm/dt/hui.svg)](https://hex.pm/packages/hui) [![License](https://img.shields.io/hexpm/l/hui.svg)](https://github.com/boonious/hui/blob/master/LICENSE) [![Last Updated](https://img.shields.io/github/last-commit/boonious/hui.svg)](https://github.com/boonious/hui/commits/master)

Hui 辉 ("shine" in Chinese) is a [Solr](http://lucene.apache.org/solr/) client and library for Elixir.

## Usage

Hui enables [Solr](http://lucene.apache.org/solr/) querying, updating and other forms of interaction
in [Elixir](https://elixir-lang.org) or [Phoenix](https://phoenixframework.org) applications.

### Example - searching

```elixir
  import Hui

  # query structs
  alias Hui.Query.{Standard,DisMax,Common,Facet,FacetRange,Suggest,MoreLikeThis,Highlight}

  url = "http://localhost:8983/solr/gettingstarted/select"
  search(url, [%Standard{q: "author:I*"}, %Facet{field: ["cat", "author_str"], mincount: 1}])

  suggester_url = "http://localhost:8983/solr/collection/suggest"
  suggest_query = %Suggest{q: "ha", count: 10, dictionary: ["name_infix", "ln_prefix", "fn_prefix"]}
  suggest(suggester_url, suggest_query)

  # DisMax SolrCloud query
  x = %DisMax{q: "market", qf: "description^2.3 title", mm: "2<-25% 9<-3", pf: "title", ps: 1, qs: 3}
  y = %Common{collection: "library,commons", rows: 10, distrib: true, "shards.tolerant": true, "shards.info": true}
  z = %Facet{field: ["cat", "author_str"], mincount: 1}
  search(url, [x, y, z])

  # more elaborated faceting query
  range1 = %FacetRange{range: "price", start: 0, end: 100, gap: 10, per_field: true}
  range2 = %FacetRange{range: "popularity", start: 0, end: 5, gap: 1, per_field: true}
  z = %Facet{field: ["cat", "author_str"], mincount: 1, range: [range1, range2]}
  search(url, [x, y, z])

  # the above spawns a request with the following query string
  #
  # q=...&
  # f.price.facet.range.end=100&
  # f.price.facet.range.gap=10&facet.range=price&
  # f.price.facet.range.start=0&
  # f.popularity.facet.range.end=5&
  # f.popularity.facet.range.gap=1&
  # facet.range=popularity&
  # f.popularity.facet.range.start=0

  # convenience functions
  suggest(suggester_url, "ha", 5, ["name_infix", "ln_prefix", "fn_prefix"], "1939")
```

Query parameters could be any arbitrary keywords such as `[q: "*", rows: 10]` or
built-in query [structs](https://hexdocs.pm/hui/Hui.html#t:solr_struct/0) that
provide a structured way for using the comprehensive and powerful features of Solr.

See the [API reference](https://hexdocs.pm/hui/api-reference.html#content)
and [Solr reference guide](http://lucene.apache.org/solr/guide/searching.html)
for more details on available search parameters.

### Example - updating

To add, update and delete Solr documents, as well as optimised search indexes:

```elixir
  import Hui

  # Specify an update handler endpoint for JSON-formatted update
  headers = [{"content-type", "application/json"}]
  url = {"http://localhost:8983/solr/collection/update", headers}

  # Solr documents
  doc1 = %{
    "actors" => ["Ingrid Bergman", "Liv Ullmann", "Lena Nyman", "Halvar Björk"],
    "desc" => "A married daughter who longs for her mother's love is visited by the latter, a successful concert pianist.",
    "directed_by" => ["Ingmar Bergman"],
    "genre" => ["Drama", "Music"],
    "id" => "tt0077711",
    "initial_release_date" => "1978-10-08",
    "name" => "Autumn Sonata"
  }
  doc2 = %{
    "actors" => ["Bibi Andersson", "Liv Ullmann", "Margaretha Krook"],
    "desc" => "A nurse is put in charge of a mute actress and finds that their personas are melding together.",
    "directed_by" => ["Ingmar Bergman"],
    "genre" => ["Drama", "Thriller"],
    "id" => "tt0060827",
    "initial_release_date" => "1967-09-21",
    "name" => "Persona"
  }

  # Add the docs and commit them to the index immediately
  update(url, [doc1, doc2])

  # Send documents to another pre-configured endpoint
  update(:updater, [doc1, doc2])

  delete_by_id(url, "tt0077711") # delete one doc
  delete_by_id(url, ["tt0077711", "tt0060827"]) # delete a list of docs
  delete_by_query(url, ["genre:Drama", "name:Persona"]) # delete via filter queries
```

Advanced update commands may be issued using the
[`Hui.Query.Update`](https://hexdocs.pm/hui/Hui.Query.Update.html) struct, as well as through
any valid binary data encapsulating Solr documents and commands.

```elixir
  import Hui
  alias Hui.Query.Update

  x = %Update{doc: [doc1, doc2], commit: true, waitSearcher: true}
  update(url, x)

  # Commits docs within 5 seconds
  x = %Update{doc: [doc1, doc2], commitWithin: 5000, overwrite: true}
  update(url, x)

  # Commit and optimise index
  update(url, %Update{commit: true, waitSearcher: true, optimize: true, maxSegments: 10})

  # Binary mode, e.g. delete a document via XML binary
  {url: "http://localhost:8983/solr/collection", [{"content-type", "application/xml"}]}
  |> update("<delete><id>9780141981727</id></delete>")
```

See [Solr reference](http://lucene.apache.org/solr/guide/uploading-data-with-index-handlers.html)
for more details on update commands, data types and formats.

### Parsed Solr response

Solr results is returned encapsulated in the `HTTP` response struct containing the Solr response.

```elixir
  {:ok,
   %Hui.Http{
    body: "...[Solr response]..",
    headers: [
      {"Content-Type", "application/json;charset=utf-8"},
      {"Content-Length", "4005"}
    ],
    method: :get,
    options: [],
    status: 200,
    url: "http://localhost:8983/solr/gettingstarted/select?q=%2A"
   }
  }
```

JSON response is automatically parsed and decoded as map.
It is accessible via the `body` key.

```elixir
  {:ok, resp} = Hui.search(solr_params)

  # getting a list of Solr documents (Map)
  solr_docs = resp.body["response"]["docs"]
  total_hits = resp.body["response"]["numFound"]
```

**Note**: other response formats such as XML, are currently being returned as raw text.

## HTTP clients and endpoints settings
Solr endpoints and request handlers may be specified in multiple ways:

```elixir
  # binary string endpoint
  Hui.search("http://localhost:8983/solr/collection/select", q: "loch")

  # atomic key referring to an endpoint in configuration - see "Configuration"
  Hui.search(:library, q: "edinburgh", rows: 10)

  # a tuple containing HTTP headers and client options
  headers = [{"accept", "application/json"}]
  options = [timeout: 10000]
  url = {"http://localhost:8983/solr/collection/suggest", headers, options}

  Hui.search(url, %Hui.Query.Suggest{q: "el", dictionary: "mySuggester"})
```

HTTP headers and client options for a specific endpoint may also be
included in the a `{url, headers, options}` tuple where:

  - `url` is the Solr HTTP endpoint including any request handler
  - `headers`: a tuple list of [HTTP headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers)
  - `options`: a keyword list of HTTP client options such as [Erlang httpc options](https://erlang.org/doc/man/httpc.html#request-4) or [HTTPoison options](https://hexdocs.pm/httpoison/HTTPoison.Request.html), e.g. `timeout`, `recv_timeout`, `max_redirect`

If `HTTPoison` is used, advanced HTTP options such as [the use of connection pools](https://github.com/edgurgel/httpoison#connection-pools)
may also be specified via `options`, e.g. `[hackney: [pool: :default]]`

### Deploying an HTTP client

Hui relies on an existing client to facilitate HTTP requests. It currently provides support
for the following clients:
 - [Erlang httpc](https://erlang.org/doc/man/httpc.html) - built-in
 - [HTTPoison](https://github.com/edgurgel/httpoison)
 - [Finch](https://github.com/sneako/finch)

Instead of using the default httpc client, one of the above HTTP clients may be deployed
by adding the client as dependency and specificying it via the `http_client` configuration - see below. 
Other HTTP clients may also be used by implementing the `Hui.Http.Client` behaviour.

```ex
  config :hui,
    http_client: Hui.Http.Clients.Finch
```

```ex
  defp deps do
    [
      ..
      {:finch, "~> 0.16"}
      ..
    ]
  end
```

For `Finch` (only), you also need to [start it from your supervision tree](https://github.com/sneako/finch#usage) and 
configure it as below:
  
```ex
  # use the same name specified in the supervision tree
  config :hui, :finch, name: FinchSolr
```

See `Hui.Http.Client` for further details.

## Installation

Hui is [available in Hex](https://hex.pm/packages/hui), the package can be installed
by adding `hui` to your list of dependencies in `mix.exs`:

```elixir
  def deps do
    [
      {:hui, "~> 0.11.0"}
    ]
  end
```

Then run `$ mix deps.get`.

Documentation can be found at [https://hexdocs.pm/hui](https://hexdocs.pm/hui).

## Configuration

A default Solr endpoint may be specified in the application configuration. HTTP headers and options may also be configured.

```elixir
  config :hui, :default,
    url: "http://localhost:8983/solr/gettingstarted/select",
    headers: [{"accept", "application/json"}]
    options: [timeout: 10_000]
```

The URL endpoint should also include
[various request handlers](http://lucene.apache.org/solr/guide/7_4/overview-of-searching-in-solr.html#overview-of-searching-in-solr)
for many purposes (search, autosuggest, spellcheck, indexing etc.). The handlers are configured
in different custom or normative names in
[Solr configuration](http://lucene.apache.org/solr/guide/7_4/requesthandlers-and-searchcomponents-in-solrconfig.html#requesthandlers-and-searchcomponents-in-solrconfig),
e.g. "select" for search queries.

Additional endpoints and request handlers can be configured in Hui using arbitrary config keys (e.g. `:suggester`):

```elixir
  config :hui, :suggester,
    url: "http://localhost:8983/solr/collection/suggest"
```

Use the config key to represent Solr endpoints in functions such as `Hui.search/3`.

## Software library

Hui [modules and data structures](https://hexdocs.pm/hui/api-reference.html#content) can be used for building Solr
application in Elixir and Phoenix. The query struct modules provide idiomatic and structured ways for
creating and encoding Solr parameters. For example, multiple filters and facet fields can be specified via list.
`fq: ["field1", "field2"]`, `field: ["field1", "field2"]`, `gap: 10` Elixir codes.

"Per-field" faceting can be specified in a succinct and unified
way, e.g. `gap` instead of the long-winded `f.[fieldname].facet.range.gap` (per field) or `facet.range.gap`
(single field). Per-field usage for a particular facet can be set or unset via the `per_field` key (example below).

`Hui.Encoder` protocol and `Hui.Encode` utility provide support for encoding query structs into binary and [IO data](#io-data-encoding) formats. A custom query struct may be developed by implementing the Encoder protocol.

```elixir
  alias Hui.Query.{Facet,FacetRange}

  %Facet{field: ["type", "year", "subject"], query: "edited:true"}
  |> Hui.Encoder.encode
  # -> "facet=true&facet.field=type&facet.field=year&facet.field=subject&facet.query=edited%3Atrue"
  # facet=true, facet prefixes are generated implicitly

  # a unified way to specify per-field or singe-field faceting
  x = %FacetRange{range: "age", gap: 10, start: 0, end: 100}
  x |> Hui.Encoder.encode
  # -> "facet.range.end=100&facet.range.gap=10&facet.range=age&facet.range.start=0"

 %{x | per_field: true} # toggle per field faceting
  |> Hui.Encoder.encode
  # -> "f.age.facet.range.end=100&f.age.facet.range.gap=10&facet.range=age&f.age.facet.range.start=0"
```

`Hui.Query.Update` struct enables
various JSON-formatted update and grouped commands to be generated.

```elixir
  alias Hui.Query.Update
  alias Hui.Encoder

  # doc1, doc2 are Maps of Solr documents
  x = %Update{doc: [doc1, doc2], commit: true, commitWithin: 1000}
  x |> Encoder.encode
  # -> "{\"add\":{\"commitWithin\":1000,\"doc\":{...}},\"add\":{\"commitWithin\":1000,\"doc\":{...}},\"commit\":{}}"

  # Delete the documents by ID
  %Update{delete_id: ["tt1316540", "tt1650453"]} |> Encoder.encode
  # -> "{\"delete\":{\"id\":\"tt1316540\"},\"delete\":{\"id\":\"tt1650453\"}}"
```

The structs and their associated type spec also provide binding to and introspection of available Solr parameters.

```elixir
  iex> %Hui.Query.Facet{field: ["type", "year"], query: "year:[2000 TO NOW]"}
  %Hui.Query.Facet{
    contains: nil,
    "contains.ignoreCase": nil,
    "enum.cache.minDf": nil,
    excludeTerms: nil,
    exists: nil,
    facet: true,
    field: ["type", "year"],
    interval: nil,
    limit: nil,
    matches: nil,
    method: nil,
    mincount: nil,
    missing: nil,
    offset: nil,
    "overrequest.count": nil,
    "overrequest.ratio": nil,
    pivot: [],
    "pivot.mincount": nil,
    prefix: nil,
    query: "year:[2000 TO NOW]",
    range: nil,
    sort: nil,
    threads: nil
  }
```

### IO data encoding
To leverage Erlang runtime and some HTTP client features for lower memory
usage and increased performance, `Hui.Encoder` provides functions to return either
string or [IO data](https://hexdocs.pm/elixir/IO.html#module-io-data)
which can be sent directly to IO functions or over the socket.

## Copyright and License

Copyright (c) 2023 Boon Low

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at [https://www.apache.org/licenses/LICENSE-2.0](https://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.11.0 (2023-04-27)
* HTTP client behaviour rewrite, adding `handle_response` callback
* introduce a response parser behaviour and implementation for parsing JSON response
* provide support for using Finch HTTP client
* refactoring in accordance with SOLID
* add integration tests and Github action workflow for all 3 supported HTTP clients: httpc, HTTPoison, Finch
* other: use Hammox/Mox in unit tests, credo linter

## 0.10.5 (2021-05-07)
* support Solr metrics API querying
* implement ping functions
* use {url, headers, options} tuple to represent endpoints, mark URL module for deprecation
* additional `handler`, `collection` fields for endpoints configuration

## 0.10.4 (2020-11-07)
* Default HTTP client behaviour implementation based on Erlang httpc
* `HTTPoison` becomes optional dependency and a configurable HTTP client

## 0.10.3 (2020-09-17)
* Rewrite `Hui.Encode.encode` utility with body-recursive functions for better performance
* Provide a new `encode_to_iodata/1` function and implementations in `Hui.Encoder` protocol

## 0.10.2 (2020-08-06)
* Extensive test refactoring

## 0.10.1 (2020-07-17)
* Simpler Http client behaviour with a single `dispatch` callback, update HTTPoison (reference) implementation
* Hui.Query module deprecation
* Refactoring and more HTTPoison dependency decoupling

## 0.10.0 (2020-07-06)
* New HTTP client architecture: introduce a behaviour with a HTTPoison implementation
* Begin HTTPoison dependency decoupling
* Removal of all bang (!) functions towards a simpler core library

## 0.9.3 (2020-05-08)
* Remove various deprecated functions and data structs
* `new` instantiation function for data structs

## 0.9.2 (2019-11-29)

* New JSON Encoder protocol implementation for `Hui.Query.Update` struct
* Refactor encoding utiliy functions: prep for forthcoming support of JSON request/facet APIs

## 0.9.1 (2019-11-19)

* New `Hui.Query.Update` struct for indexing documents
* `Hui.Encoder` encodes update query in JSON format
* Hui.Query.get/2 and Hui.Query.post/2 functions (accept built-in data structs) for HTTP requests

## 0.9.0 (2019-11-06)

* New query structs: `Hui.Query.Standard`, `Hui.Query.DisMax`, `Hui.Query.Common`, Faceting: `Hui.Query.Facet`, `Hui.Query.FacetRange`, `Hui.Query.FacetInterval`, `Hui.Query.Highlight`, `Hui.Query.HighlighterFastVector`, `Hui.Query.HighlighterOriginal`, `Hui.Query.HighlighterUnified`, `Hui.Query.SpellCheck`, `Hui.Query.Suggest` `Hui.Query.MoreLikeThis`
* `Hui.Encoder` protocol and implementation for encoding the new query structs

## 0.8.3, 0.8.4 (2019-10-03)

* update dependencies: newer versions of bypass, httpoison, cowboy

## 0.8.2 (2018-10-09)

* `update/3`, `update!/3` functions for adding and updating Solr docs
* `delete/3`, `delete!/3`, `delete_by_query/3`, `delete_by_query!/3` functions for deleting Solr docs
* `commit/2`, `commit!/2` functions for committing Solr docs to index

## 0.8.1 (2018-10-08)

* Documentation, README and doctests for Solr updating via `Hui.U` struct

## 0.8.0 (2018-10-06)

* Solr updating via struct: `Request.update/3` now accepts a `Hui.U` struct for adding, deleting and committing documents, as well as optimising indexes

## 0.7.0 (2018-10-03)

* Enable Solr updating with binary data (JSON, XML update docs) via `Request.update/3` function

## 0.6.3 (2018-09-28)

* Convenience functions for search and suggester requests (`q/6`, `search/7`, `suggest/5`)
* Bangified all main module functions (`q!/1`, `search!/2`, `spellcheck!/2`, `spellcheck!/3`, `suggest!/2`, `mlt!/3`)
* Deprecate the standard query + faceting functions (`q/2`, `search/3`) as these are now provided through the more general-purpose functions (`q/1`, `search/2`)

## 0.6.2 (2018-09-26)

* Introduce a "bang" feature via an additional boolean parameter in `Request.search/3` for implementing bangified functions later
* Deprecate `Hui.Search` module which is replaced by `Hui.Request` to provide update functions later
* Refactor search tests with a helper function that tests `search/2` in both `Hui` and `Hui.Request` modules
* Refactor and fix typespec issues with Dialyzir
* Code coverage metric via ExCoveralls and coveralls.io

## 0.6.1 (2018-09-24)

* Provide `cursorMark` parameter for deep paging in `Hui.Q`
* Define and use a custom exception `Hui.Error`, instead of the mixed usage of HTTPoison and arbitrary error text responses

## 0.6.0 (2018-09-21)

* Support SolrCloud request parameters such as 'collection', 'shards', 'shards.tolerant' in `Hui.Q` and distributed search

## 0.5.7 (2018-09-21)

* Support autosuggest (suggester) query via `suggest/2`
* Support spell checking query via `spellcheck/3`
* Support MoreLikeThis query via `mlt/3`
* Suggester query via `Hui.S` struct
* Spell checking query via `Hui.Sp` struct
* MoreLikeThis query via `Hui.M` struct

## 0.5.6 (2018-09-20)

* Results highlighting (snippets) query via `Hui.H` struct
* Unified, Original, Fast Vector results highlighting (snippets) via `Hui.H1/H2/H3` structs

## 0.5.5 (2018-09-20)

* `q/1`, `search/2` to generally accept a list of current and forthcoming Hui structs for more complex querying, e.g. DisMax
* DisMax (Maximum Disjunction) and Extended DisMax (eDisMax) query now via a new `Hui.D` struct
* Deprecate DisMax parameters from `Hui.Q` struct

## 0.5.4 (2018-09-16)

* Maximum Disjunction (DisMax) query via `Hui.Q` struct

## 0.5.3 (2018-09-14)

* Support JSON and XML response writer parameters in `Hui.Q` struct

## 0.5.2 (2018-09-13)

* New `q/2`, `search/3` features for structured querying and faceting - Solr standard parser
* `search/2`, `q/1` accepts query struct (Hui.Q) parameters
* `Search.search/2` works with query and faceting structs (`Hui.Q`, `Hui.F`) parameters

## 0.5.1 (2018-09-12)


lib/hui.ex
defmodule Hui do
  @moduledoc """
  Hui 辉 ("shine" in Chinese) is an [Elixir](https://elixir-lang.org) client and library for
  [Solr enterprise search platform](http://lucene.apache.org/solr/).

  ### Usage

  - Searching Solr: `search/3`
  - Updating: `update/3`, `delete/3`, `delete_by_query/3`, `commit/2`
  - Other: `suggest/2`, `suggest/5`
  - Admin: `metrics/2`, `ping/1`
  - [README](https://hexdocs.pm/hui/readme.html#usage)
  """

  alias Hui.Http
  alias Hui.Query

  @client Hui.Http.Client.impl()

  @type endpoint :: Http.endpoint()
  @type query :: Http.query()
  @type update_query :: Http.update_query()

  @type http_response :: Http.response()

  @doc """
  Issue a keyword list or structured query to a Solr endpoint.

  ### Example - parameters

  ```
    url = "http://localhost:8983/solr/collection"

    # a keyword list of arbitrary parameters
    Hui.search(url, q: "edinburgh", rows: 10)

    # supply a list of Hui structs for more complex query e.g. DisMax
    alias Hui.Query

    x = %Query.DisMax{q: "run", qf: "description^2.3 title", mm: "2<-25% 9<-3"}
    y = %Query.Common{rows: 10, start: 10, fq: ["edited:true"]}
    z = %Query.Facet{field: ["cat", "author_str"], mincount: 1}
    Hui.search(url, [x, y, z])

    # SolrCloud query
    x = %Query.DisMax{q: "john"}
    y = %Query.Common{collection: "library,commons", rows: 10, distrib: true, "shards.tolerant": true, "shards.info": true}
    Hui.search(url, [x,y])

    # With results highlighting (snippets)
    x = %Query.Standard{q: "features:photo"}
    y = %Query.Highlight{fl: "features", usePhraseHighlighter: true, fragsize: 250, snippets: 3}
    Hui.search(url, [x, y])
  ```

  ### Example - faceting

  ```
    alias Hui.Query

    range1 = %Query.FacetRange{range: "price", start: 0, end: 100, gap: 10, per_field: true}
    range2 = %Query.FacetRange{range: "popularity", start: 0, end: 5, gap: 1, per_field: true}

    x = %Query.DisMax{q: "ivan"}
    y = %Query.Facet{field: ["cat", "author_str"], mincount: 1, range: [range1, range2]}

    Hui.search(:default, [x, y])
  ```

  The above `Hui.search(:default, [x, y])` example issues a request that resulted in
  the following Solr response header showing the corresponding generated and encoded parameters.

  ```json
  "responseHeader" => %{
    "QTime" => 106,
    "params" => %{
      "f.popularity.facet.range.end" => "5",
      "f.popularity.facet.range.gap" => "1",
      "f.popularity.facet.range.start" => "0",
      "f.price.facet.range.end" => "100",
      "f.price.facet.range.gap" => "10",
      "f.price.facet.range.start" => "0",
      "facet" => "true",
      "facet.field" => ["cat", "author_str"],
      "facet.mincount" => "1",
      "facet.range" => ["price", "popularity"],
      "q" => "ivan"
    },
    "status" => 0,
    "zkConnected" => true
  }
  ```

  ### URLs, Headers, Options

  HTTP headers and client options for a specific endpoint may also be
  included in the a `{url, headers, options}` tuple where:

  - `url` is a typical Solr endpoint that includes a request handler
  - `headers`: a tuple list of [HTTP headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers)
  - `options`: a keyword list of configured http client options such as [Erlang httpc](https://erlang.org/doc/man/httpc.html#request-4),
   [HTTPoison](https://hexdocs.pm/httpoison/HTTPoison.Request.html), e.g.
  `timeout`, `recv_timeout`, `max_redirect`

  If `HTTPoison` is used, advanced HTTP options such as the use of connection pools
  may also be specified via `options`.
  """
  @spec search(endpoint, query, module) :: http_response
  defdelegate search(endpoint, query, client), to: Http, as: :get

  @doc """
  Issue a structured suggest query to a specified Solr endpoint.

  ### Example

  ```
    # :library is a configured endpoint
    suggest_query = %Hui.Query.Suggest{q: "ha", count: 10, dictionary: "name_infix"}
    Hui.suggest(:library, suggest_query)
  ```
  """
  @spec suggest(endpoint, Query.Suggest.t()) :: http_response
  defdelegate suggest(endpoint, query), to: Hui.Suggest

  @doc """
  Convenience function for issuing a suggester query to a specified Solr endpoint.

  ### Example

  ```
    # :autocomplete is a configured endpoint
    Hui.suggest(:autocomplete, "t")
    Hui.suggest(:autocomplete, "bo", 5, ["name_infix", "ln_prefix", "fn_prefix"], "1939")
  ```
  """
  @spec suggest(endpoint, binary, nil | integer, nil | binary | list(binary), nil | binary) :: http_response
  defdelegate suggest(endpoint, q, count, dictionaries, context), to: Hui.Suggest

  @doc """
  Updates or adds Solr documents to an index or collection.

  This function accepts documents as map (single or a list) and commits the docs
  to the index immediately by default - set `commit` to `false` for manual or
  auto commits later.

  It can also operate in update struct and binary modes,
  the former uses the `t:Hui.Query.Update.t/0` struct
  while the latter acepts text containing any valid Solr update data or commands.

  An index/update handler endpoint should be specified through a URL string or
  {url, headers, options} tuple for headers and HTTP client options specification.

  A "content-type" request header is required so that Solr knows the
  incoming data format (JSON, XML etc.) and can process data accordingly.

  ### Example - map, list and binary data

  ```
    # Index handler for JSON-formatted update
    headers = [{"content-type", "application/json"}]
    endpoint = {"http://localhost:8983/solr/collection/update", headers}

    # Solr docs in maps
    doc1 = %{
      "actors" => ["Ingrid Bergman", "Liv Ullmann", "Lena Nyman", "Halvar Björk"],
      "desc" => "A married daughter who longs for her mother's love is visited by the latter, a successful concert pianist.",
      "directed_by" => ["Ingmar Bergman"],
      "genre" => ["Drama", "Music"],
      "id" => "tt0077711",
      "initial_release_date" => "1978-10-08",
      "name" => "Autumn Sonata"
    }

    doc2 = %{
      "actors" => ["Bibi Andersson", "Liv Ullmann", "Margaretha Krook"],
      "desc" => "A nurse is put in charge of a mute actress and finds that their personas are melding together.",
      "directed_by" => ["Ingmar Bergman"],
      "genre" => ["Drama", "Thriller"],
      "id" => "tt0060827",
      "initial_release_date" => "1967-09-21",
      "name" => "Persona"
    }

    Hui.update(endpoint, doc1) # add a single doc
    Hui.update(endpoint, [doc1, doc2]) # add a list of docs

    # Don't commit the docs e.g. mass ingestion when index handler is setup for autocommit.
    Hui.update(endpoint, [doc1, doc2], false)

    # Send to a configured endpoint
    Hui.update(:updater, [doc1, doc2])

    # Binary mode, add and commit a doc
    Hui.update(endpoint, "{\\\"add\\\":{\\\"doc\\\":{\\\"name\\\":\\\"Blade Runner\\\",\\\"id\\\":\\\"tt0083658\\\",..}},\\\"commit\\\":{}}")

    # Binary mode, delete a doc via XML
    headers = [{"content-type", "application/xml"}]
    endpoint = {"http://localhost:8983/solr/collection/update", headers}
    Hui.update(endpoint, "<delete><id>9780141981727</id></delete>")

  ```

  ### Example - `t:Hui.Query.Update.t/0` and other update options
  ```

    # endpoint, doc1, doc2 from the above example
    ...

    # Hui.Query.Update struct command for updating and committing the docs to Solr within 5 seconds

    alias Hui.Query

    x = %Query.Update{doc: [doc1, doc2], commitWithin: 5000, overwrite: true}
    {status, resp} = Hui.update(endpoint, x)

    # Delete the docs by IDs, with a URL key from configuration
    {status, resp} = Hui.update(:library_update, %Query.Update{delete_id: ["tt1316540", "tt1650453"]})

    # Commit and optimise index, keep max index segments at 10
    {status, resp} = Hui.update(endpoint, %Query.Update{commit: true, waitSearcher: true, optimize: true, maxSegments: 10})

    # Commit index, expunge deleted docs
    {status, resp} = Hui.update(endpoint, %Query.Update{commit: true, expungeDeletes: true})
  ```
  """
  @spec update(endpoint, update_query) :: http_response
  defdelegate update(endpoint, query), to: Http, as: :post

  @spec update(endpoint, update_query, boolean, module) :: http_response
  defdelegate update(endpoint, query, commit, client), to: Http, as: :post

  @doc """
  Deletes Solr documents.

  This function accepts a single or list of IDs and immediately delete the corresponding
  documents from the Solr index (commit by default).

  An index/update handler endpoint should be specified through a URL string
  or {url, headers, options} tuple.

  A JSON "content-type" request header is required so that Solr knows the
  incoming data format and can process data accordingly.

  ### Example
  ```
    # Index handler for JSON-formatted update
    headers = [{"content-type", "application/json"}]
    endpoint = {"http://localhost:8983/solr/collection/update", headers}

    Hui.delete_by_id(endpoint, "tt2358891") # delete a single doc
    Hui.delete_by_id(endpoint, ["tt2358891", "tt1602620"]) # delete a list of docs

    Hui.delete_by_id(endpoint, ["tt2358891", "tt1602620"], false) # delete without immediate commit
  ```
  """
  @spec delete_by_id(endpoint, binary | list(binary), boolean, module) :: http_response
  def delete_by_id(endpoint, ids, commit \\ true, client \\ @client) when is_binary(ids) or is_list(ids) do
    Http.post(endpoint, %Query.Update{delete_id: ids, commit: commit}, commit, client)
  end

  # coveralls-ignore-start
  @deprecated "Use delete_by_id/3 instead"
  def delete(endpoint, ids, commit \\ true) when is_binary(ids) or is_list(ids) do
    Http.post(endpoint, %Query.Update{delete_id: ids, commit: commit})
  end

  # coveralls-ignore-stop

  @doc """
  Deletes Solr documents by filter queries.

  This function accepts a single or list of filter queries and immediately delete the corresponding
  documents from the Solr index (commit by default).

  An index/update handler endpoint should be specified through a URL string
  or {url, headers, options} tuple.

  A JSON "content-type" request header is required so that Solr knows the
  incoming data format and can process data accordingly.

  ### Example
  ```
    # Index handler for JSON-formatted update
    headers = [{"content-type", "application/json"}]
    endpoint = {"http://localhost:8983/solr/collection", headers}

    Hui.delete_by_query(endpoint, "name:Persona") # delete with a single filter
    Hui.delete_by_query(endpoint, ["genre:Drama", "name:Persona"]) # delete with a list of filters
  ```
  """
  @spec delete_by_query(endpoint, binary | list(binary), boolean) :: http_response
  def delete_by_query(endpoint, q, commit \\ true, client \\ @client) when is_binary(q) or is_list(q) do
    Http.post(endpoint, %Query.Update{delete_query: q, commit: commit}, commit, client)
  end

  @doc """
  Commit any added or deleted Solr documents to the index.

  This provides a (separate) mechanism to commit previously added or deleted documents to
  Solr index for different updating and index maintenance scenarios. By default, the commit
  waits for a new Solr searcher to be regenerated, so that the commit result is made available
  for search.

  An index/update handler endpoint should be specified through a URL string
  or {url, headers, options} tuple.

  A JSON "content-type" request header is required so that Solr knows the
  incoming data format and can process data accordingly.

  ### Example
  ```
    # Index handler for JSON-formatted update
    headers = [{"content-type", "application/json"}]
    endpoint = {"http://localhost:8983/solr/collection", headers}

    Hui.commit(endpoint) # commits, make new docs available for search
    Hui.commit(endpoint, false) # commits op only, new docs to be made available later
  ```

  Use `t:Hui.Query.Update.t/0` struct for other types of commit and index optimisation, e.g. expunge deleted docs to
  physically remove docs from the index, which could be a system-intensive operation.
  """
  @spec commit(endpoint, boolean) :: http_response
  def commit(endpoint, wait_searcher \\ true, client \\ @client) do
    Http.post(endpoint, %Query.Update{commit: true, waitSearcher: wait_searcher}, true, client)
  end

  @doc """
  Retrieves metrics data from the Solr admin API.

  ### Example
  ```
    endpoint = {"http://localhost:8983/solr/admin/metrics", [{"content-type", "application/json"}]}
    Hui.metrics(endpoint, group: "core", type: "timer", property: ["mean_ms", "max_ms", "p99_ms"])
  ```
  """
  @spec metrics(endpoint, keyword) :: http_response
  defdelegate metrics(endpoint, options), to: Hui.Admin

  @doc """
  Ping a given endpoint.

  ### Example
  ```
    # ping a configured atomic endpoint
    Hui.ping(:gettingstarted)

    # directly ping a binary URL
    Hui.ping("http://localhost:8983/solr/gettingstarted/admin/ping")
  ```

  Successful ping returns a `{:pong, qtime}` tuple, whereas failure gets a `:pang` response.
  """
  @spec ping(binary() | atom()) :: {:pong, integer} | :pang
  defdelegate ping(endpoint), to: Hui.Admin

  @doc """
  Ping a given endpoint with options.

  Raw HTTP response is returned when options such as `wt`, `distrib` is provided:
  ```
    Hui.ping(:gettingstarted, wt: "xml", distrib: false)
    # -> returns {:ok, %Hui.HTTP{body: "raw HTTP response", status: 200, ..}}
  ```
  """
  @spec ping(binary() | atom(), keyword) :: http_response
  defdelegate ping(endpoint, options), to: Hui.Admin
end
